blaze/expr/collections.py

```diff
@@ -31,19 +33,200 @@
+        return "%s.%s(%s)" % (self._child, func, key)
+
+
+def _isrecord(obj):
+    return isinstance(obj, Expr) and isrecord(obj.dshape.measure)
+
+
+def _find_table_and_field(obj):
+    while True:
+        try:
+            child = obj._child
+        except AttributeError:
+            return None, None
+        if isinstance(obj, Field) and _isrecord(child):
+            return child, obj
+        obj = child
+
+
+def _extract_key_args(first, second, func):
+    child = first
+    if isinstance(first, _strtypes) and second is None: # asc('a')
+         child = None
+         table = None
+         key = (first,)
+    elif _isrecord(first):
+         table = first
+         if second is None: # asc(data)
+             key = tuple(table.dshape.measure.names)
+         elif isinstance(second, _strtypes) and second in table.dshape.measure.names: # asc(data, 'a')
+             key = (second,)
+         elif hasattr(second, '_child') and second._child.isidentical(table): # asc(data, data.a)
+             key = (second._name,)
+         else:
+             raise ValueError("%s(): invalid key for table %s" % (func, str(table)))
+    else:
+         table, field = _find_table_and_field(first)
+         if table is None:
+             raise ValueError("%s(): invalid key for table %s" % (func, str(table)))
+
+         if second is None: # asc(data.a)
+             key = (field._name,)
+         elif isinstance(second, _strtypes) and second in table.dshape.measure.names: # asc(data.a, 'a')
+             key = (second,)
+         elif hasattr(second, '_child') and second._child.isidentical(table): # asc(data.a, data.a)
+             key = (second._name,)
+         else:
+             raise ValueError("%s(): invalid key for table %s" % (func, str(table)))
+    return child, table, key
+
+
+def asc(first=None, second=None):
+    """Sort an expression in ascending order"""
+    child, table, key = _extract_key_args(first, second, 'asc')
+    return Key(child, table, key, True)
+
+
+def desc(first=None, second=None):
+    """Sort an expression in descending order"""
+    child, table, key = _extract_key_args(first, second, 'desc')
+    return Key(child, table, key, False)
+
+
+class Sort2(Expr):
+
+    """ Table in sorted order
+
+    Examples
+    --------
+    >>> import blaze as bz
+    >>> from blaze import symbol
+    >>> accounts = symbol('accounts', 'var * {name: string, amount: int}')
+    >>> accounts.sort2(bz.desc('amount')).schema
+    dshape("{name: string, amount: int32}")
+    """
+
+    _arguments = '_child', '_keys'
+
+    def _dshape(self):
+        return self._child.dshape
+
+    @property
+    def keys(self):
+        return list(self._keys)
+
+    def _len(self):
+        return self._child._len()
+
+    @property
+    def _name(self):
+        return self._child._name
+
+    def __str__(self):
+        return "%s.sort2(%s)" % (self._child, ", ".join(str(x) for x in self.keys))
+
+
+def sort2(first, *rest):
```

Comment:
I don't have a strong opinion on the name -- eventually the current `sort` will be replaced with this new thing, so in the long run, we'll only have `sort`.

+1 on raising a deprecation warning in current `sort`.